tests/test_upload.py

```diff
@@ -444,3 +447,73 @@ def test_upload_registration_fail(rse, scope, upload_client_registration_fail, f
             'did_name': fn,
             'guid': generate_uuid()
         }])
+
+
+@pytest.mark.parametrize("structure,failure", [
+    ({"container": ["file1", "file2", {"dataset": ["file3", "file4"]}]}, True),  # Cannot have both files and datasets in a container
+    ({
+        "container": {
+            "dataset": ["file1", "file2"]
+            },
+        "container2": {
+            "dataset": ["file1", {"dataset2": ["file1"]}]
+        }
+    }, True),  # One valid container, another invalid container
```

Comment:
I would name the invalid container `bad_container` or something similar to clarify which one is the invalid one (and maybe also the naming the valid containers as `good_container` / `valid_container` or something along those lines)